src/pages/images/actions/forms/UploadImageForm.tsx

```diff
@@ -0,0 +1,197 @@
+import { ChangeEvent, FC, useState } from "react";
+import { useEventQueue } from "context/eventQueue";
+import { useFormik } from "formik";
+import { useToastNotification } from "context/toastNotificationProvider";
+import { createImageAlias, uploadImage } from "api/images";
+import {
+  ActionButton,
+  Button,
+  Form,
+  Input,
+  Modal,
+} from "@canonical/react-components";
+import * as Yup from "yup";
+import { Link } from "react-router-dom";
+import { humanFileSize } from "util/helpers";
+import ProgressBar from "components/ProgressBar";
+import { UploadState } from "types/storage";
+import Loader from "components/Loader";
+import { useQueryClient } from "@tanstack/react-query";
+import { queryKeys } from "util/queryKeys";
+import { LxdSyncResponse } from "types/apiResponse";
+import { AxiosError } from "axios";
+
+interface Props {
+  close: () => void;
+  project: string;
+}
+
+const UploadImageForm: FC<Props> = ({ close, project }) => {
+  const eventQueue = useEventQueue();
+  const toastNotify = useToastNotification();
+  const [uploadState, setUploadState] = useState<UploadState | null>(null);
+  const queryClient = useQueryClient();
+
+  const notifySuccess = () => {
+    const uploaded = <Link to={`/ui/project/${project}/images`}>uploaded</Link>;
+    toastNotify.success(<>Image {uploaded}.</>);
+  };
+
+  const changeFile = (e: ChangeEvent<HTMLInputElement>) => {
+    if (e.target.files) {
+      void formik.setFieldValue("fileList", e.target.files);
+    }
+  };
+
+  const getImageUploadBody = (fileList: FileList): File | FormData => {
+    try {
+      if (fileList.length === 1) {
+        return fileList[0];
+      } else if (fileList.length > 2) {
+        throw new Error("Can only upload up to 2 files");
+      } else {
+        // Sorting by Size. The metadata file is very likely to be smaller than the image itself.
+        const formData = new FormData();
+        const sortedFiles = Array.from(fileList).sort(
+          (a, b) => a.size - b.size,
+        );
+
+        formData.append("metadata", sortedFiles[0]);
+        formData.append("rootfs.img", sortedFiles[1]);
+
+        return formData;
+      }
+    } catch (e) {
+      close();
+      toastNotify.failure(`Image upload failed`, e);
+      throw e;
+    }
```

Comment:
I think we shouldn't throw/catch in here, the check if have more than two files should be in line85 with the `if (values.fileList)` check.

Reply:
Thank you, please see my review :)